integrations/acquisition/covidcast/test_is_latest_issue.py

```diff
@@ -0,0 +1,346 @@
+              3, 3, 3, nmv, nmv, nmv, 20200414, 0)
+          ]  
+      self._db.insert_or_update_bulk(newGeoValues)
+      self._db.run_dbjobs()
+
+      self._db._cursor.execute('''SELECT `geo_type`,`geo_value` FROM `geo_dim`''')
+      record = self._db._cursor.fetchall()
+      self.assertEqual(len(list(record)),self.geoDimRows + 3) #2 + 3 new
+
+      self._db._cursor.execute('''SELECT * FROM `signal_latest`''')
+      record = self._db._cursor.fetchall()
+      self.assertEqual(len(list(record)),self.sigLatestRows + 6 + 3) #total entries = 2(initial) + 6(test)
+
+  @unittest.skip("helper function")
+  def view_table(table_name):
+    #TODO: abstract into function
+    y = PrettyTable()
+    y.field_names = ['signal_data_id',
+    'signal_key_id',
+    'geo_key_id',
+    'demog_key_id' ,
+    'issue' ,
+    'data_as_of_dt',
+    'time_type',
+    'time_value' ,
+    'reference_dt' ,
+    'value' ,
+    'stderr' ,
+    'sample_size',
+    'lag' ,
+    'value_updated_timestamp' ,
+    'computation_as_of_dt']
+   
+    print('\n')
+    self._db._cursor.execute('''SELECT * FROM `signal_latest`''')
+    record = self._db._cursor.fetchall()
+    print(record)
+    # assert(len(record) == )
+    for row in record:
+      y.add_row(list(row)[:len(y.field_names)])
+    print(y)
+    print("signal_latest updated")
+    # self._db._cursor.execute('''SELECT * FROM `signal_dim`''')
+    # record = self._db._cursor.fetchall()
+    # print(record)
+    # self._db._cursor.execute('''SELECT * FROM `geo_dim`''')
+    # record = self._db._cursor.fetchall()
+    # print(record)
+    
+    # sig_hist = PrettyTable()
+    # sig_hist.field_names = [('signal_data_id',), ('signal_key_id',), ('geo_key_id',), ('demog_key_id',), ('issue',), ('data_as_of_dt',), ('time_type',), ('time_value',), ('reference_dt',), ('value',), ('stderr',), ('sample_size',), ('lag',), ('value_updated_timestamp',), ('computation_as_of_dt',), ('is_latest_issue',), ('missing_value',), ('missing_stderr',), ('missing_sample_size',), ('legacy_id',)]
+    
+    # sig_load = PrettyTable()
+    # sig_load.field_names = [('signal_data_id',), ('signal_key_id',), ('geo_key_id',), ('demog_key_id',), ('issue',), ('data_as_of_dt',), ('source',), ('signal',), ('geo_type',), ('geo_value',), ('time_type',), ('time_value',), ('reference_dt',), ('value',), ('stderr',), ('sample_size',), ('lag',), ('value_updated_timestamp',), ('computation_as_of_dt',), ('is_latest_issue',), ('missing_value',), ('missing_stderr',), ('missing_sample_size',), ('legacy_id',), ('compressed_signal_key',), ('compressed_geo_key',), ('compressed_demog_key',), ('process_status',)]
+    
+    #when uploading data patches (data in signal load has < issue than data in signal_latest)
+    #when signal_load is older than signal_latest, we patch old data (i.e changed some old entries)
+    pass
+
+  @unittest.skip("Having different (time_value,issue) pairs in one call to db pipeline does not happen in practice")
```

Comment:
Can one of you clarify for me? I thought the whole point of batch issue uploads was that they included multiple issues.

Reply:
Oh! I was writing tests with multiple issues but it was not performing as intended so I sought George out for some help. Would it be possible to find a time to sit down together to go through this quickly?

Reply:
I found it -- `insert_or_update_bulk` is called from within a loop, such that it is called separately for each CSV file. Each CSV contains only one time_value and only one issue.

https://github.com/cmu-delphi/delphi-epidata/blob/61a3d8bcd8fe3bb0b2d362eee8ce5984fbc9ea4d/src/acquisition/covidcast/csv_to_database.py#L91-L109

However, this test also includes `run_dbjobs`, which is not called as part of the `csv_to_database` script or as part of `insert_or_update_bulk`. We cannot therefore expect that `run_dbjobs` will only deal with one time_value and issue at a time. So it depends on what this test is trying to verify:

* correct behavior of `insert_or_update_bulk` alone? --> then it should not call run_dbjobs
* correct behavior of `insert_or_update_bulk` and `run_dbjobs` when run separately with each CSV file? --> then this test doesn't tell us whether run_dbjobs will run correctly as part of the intended load pipeline
* correct behavior of `insert_or_update_bulk` when run separately with each CSV file, and of `run_dbjobs` when run after all available CSVs have been loaded into `signal_load`? --> then this test should run `insert_or_update_bulk` once for each time_value x issue configuration, and then `run_dbjobs` once at the end.

Reply:
It is a flaw (for some definition of 'flaw') that we cannot properly process multiple issues per datapoint in a single run of `insert_or_update[+dbjobs]` if the issues are not provided in increasing order (or really, if the 'latest' issue is not last in the series).  In practice, `insert_or_update_bulk()` is only called by `csv_to_database:upload_archive()` and each call only specifies a single issue (in fact, the only part of the composite key that differs is the `geo_value`).  This is not going to cause us problems for the foreseeable future, but it is still worth discussing to determine if/how we want to account for possibilities later (adding/changing logic in `insert_or_update` (in the SQL or in the python), carrying some of those key-field arguments further down the call stack to prevent unintended consequences, refactoring the larger ingestion pipeline, or otherwise).

Reply:
Results of zoom discussion:
Options:
- one CSV per insert, many CSVs per dbjobs
- **Decision**: one CSV per insert+dbjobs. Have insert call dbjobs itself so csv_to_database doesn't have to bother about it.
  - if someone chooses to run dbjobs on an untrusted signal_load, it's their responsibility to ensure that issues are ordered correctly.
  - could add a guard to dbjobs to check if signal_load is sorted correctly but that check would take a long time so let's not. **comment it obsessively though at the head of dbjobs.**

Options for fixing at dbjobs time: none of them are good so let's not.
- get max before loading into latest
- sort while loading into latest

Batch issue uploads do traverse issues in the correct order, but since we're having insert call dbjobs we don't need to care.